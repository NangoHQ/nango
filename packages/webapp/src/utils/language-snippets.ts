import { HTTPSnippet } from 'httpsnippet-lite';

import { indentLines, modelToString } from './scripts';

import type { NangoSyncEndpointV2 } from '@nangohq/types';
import type { TargetId } from 'httpsnippet-lite';
import type { JSONSchema7 } from 'json-schema';

function maskSecret(secret: string): string {
    return `${secret.substring(0, 4)}****`;
}

export function nodeSyncSnippet({
    modelName,
    secretKey,
    connectionId,
    providerConfigKey,
    hideSecret = true
}: {
    modelName: string;
    secretKey: string;
    connectionId: string;
    providerConfigKey: string;
    hideSecret?: boolean;
}) {
    return [
        `import { Nango } from '@nangohq/node';`,
        `const nango = new Nango({ secretKey: '${hideSecret ? maskSecret(secretKey) : secretKey}' });`,
        '',
        `const records = await nango.listRecords({`,
        `    providerConfigKey: '${providerConfigKey}',`,
        `    connectionId: '${connectionId}',`,
        `    model: '${modelName}'`,
        `});`
    ].join('\n');
}

export function nodeActionSnippet({
    actionName,
    secretKey,
    connectionId,
    providerConfigKey,
    input,
    hideSecret = true
}: {
    actionName: string;
    secretKey: string;
    connectionId: string;
    providerConfigKey: string;
    input?: JSONSchema7;
    hideSecret?: boolean;
}) {
    const triggerActionParams = [`'${providerConfigKey}'`, `'${connectionId}'`, `'${actionName}'`];

    if (input) {
        triggerActionParams.push(modelToString(input, true));
    }

    return [
        `import Nango from '@nangohq/node';`,
        `const nango = new Nango({ secretKey: '${hideSecret ? maskSecret(secretKey) : secretKey}' });`,
        `const response = await nango.triggerAction(`,
        indentLines(triggerActionParams.join(',\n')),
        `);`
    ].join('\n');
}

const languageToSpec: Record<string, string> = { javascript: 'fetch', php: 'guzzle' };
export async function httpSnippet({
    baseUrl,
    endpoint,
    secretKey,
    connectionId,
    providerConfigKey,
    language,
    input,
    hideSecret = true
}: {
    baseUrl: string;
    endpoint: NangoSyncEndpointV2;
    secretKey: string;
    connectionId: string;
    providerConfigKey: string;
    language: TargetId;
    input?: JSONSchema7 | undefined;
    hideSecret?: boolean;
}) {
    const snippet = new HTTPSnippet({
        method: endpoint.method,
        url: `${baseUrl}/v1${endpoint.path}`,
        headers: [
            { name: 'Authorization', value: `Bearer ${hideSecret ? maskSecret(secretKey) : secretKey}` },
            { name: 'Content-Type', value: 'application/json' },
            { name: 'Connection-Id', value: connectionId },
            { name: 'Provider-Config-Key', value: providerConfigKey }
        ],
        postData: input
            ? {
                  mimeType: 'application/json',
                  text: language === 'shell' ? modelToString(input, true) : modelToString(input)
              }
            : undefined,
        bodySize: 10,
        cookies: [],
        headersSize: 0,
        httpVersion: '1',
        queryString: []
    });

    return (await snippet.convert(language, languageToSpec[language], { checkErrors: true })) as string;
}
