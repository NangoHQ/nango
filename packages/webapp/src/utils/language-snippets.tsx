import type { NangoModel, NangoSyncEndpointV2 } from '@nangohq/types';
import type { TargetId } from 'httpsnippet-lite';
import { HTTPSnippet } from 'httpsnippet-lite';
import type { NangoSyncModel } from '../types';
import { modelToString } from './scripts';

function maskSecret(secret: string): string {
    return `${secret.substring(0, 4)}****`;
}

export function nodeSyncSnippet({
    modelName,
    secretKey,
    connectionId,
    providerConfigKey,
    hideSecret = true
}: {
    modelName: string;
    secretKey: string;
    connectionId: string;
    providerConfigKey: string;
    hideSecret?: boolean;
}) {
    return `import { Nango } from '@nangohq/node';
const nango = new Nango({ secretKey: '${hideSecret ? maskSecret(secretKey) : secretKey}' });

const records = await nango.listRecords({
    providerConfigKey: '${providerConfigKey}',
    connectionId: '${connectionId}',
    model: '${modelName}'
});
`;
}

export function nodeActionSnippet({
    actionName,
    secretKey,
    connectionId,
    providerConfigKey,
    input,
    hideSecret = true
}: {
    actionName: string;
    secretKey: string;
    connectionId: string;
    providerConfigKey: string;
    input?: NangoModel | NangoSyncModel;
    hideSecret?: boolean;
}) {
    let snippet = `import Nango from '@nangohq/node';
const nango = new Nango({ secretKey: '${hideSecret ? maskSecret(secretKey) : secretKey}' });

const response = await nango.triggerAction(
    '${providerConfigKey}',
    '${connectionId}',
    '${actionName}'`;
    if (input && Object.keys(input).length > 0) {
        snippet += `,
${modelToString(input)}`;
    }
    snippet += `
);`;
    return snippet;
}

const languageToSpec: Record<string, string> = { javascript: 'fetch', php: 'guzzle' };
export async function httpSnippet({
    baseUrl,
    endpoint,
    secretKey,
    connectionId,
    providerConfigKey,
    language,
    input,
    hideSecret = true
}: {
    baseUrl: string;
    endpoint: NangoSyncEndpointV2;
    secretKey: string;
    connectionId: string;
    providerConfigKey: string;
    language: TargetId;
    input?: NangoModel | NangoSyncModel | undefined;
    hideSecret?: boolean;
}) {
    const snippet = new HTTPSnippet({
        method: endpoint.method,
        url: `${baseUrl}/v1${endpoint.path}`,
        headers: [
            { name: 'Authorization', value: `Bearer ${hideSecret ? maskSecret(secretKey) : secretKey}` },
            { name: 'Content-Type', value: 'application/json' },
            { name: 'Connection-Id', value: connectionId },
            { name: 'Provider-Config-Key', value: providerConfigKey }
        ],
        postData: input
            ? { mimeType: 'application/json', text: language === 'shell' ? modelToString(input) : modelToString(input).replaceAll('\n', '') }
            : undefined,
        bodySize: 10,
        cookies: [],
        headersSize: 0,
        httpVersion: '1',
        queryString: []
    });

    return (await snippet.convert(language, languageToSpec[language], { checkErrors: true })) as string;
}
