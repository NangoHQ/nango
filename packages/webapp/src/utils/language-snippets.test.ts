import { describe, expect, it } from 'vitest';

import { httpSnippet, nodeActionSnippet, nodeSyncSnippet } from './language-snippets';

import type { NangoSyncEndpointV2 } from '@nangohq/types';
import type { JSONSchema7 } from 'json-schema';

describe('language snippets', () => {
    const sampleInput: JSONSchema7 = {
        type: 'object',
        required: ['aString', 'aNumber', 'aBoolean', 'anArrayOfStrings', 'anObject', 'anArrayOfObjects', 'anArrayOfReferencedModel', 'aUnion'],
        properties: {
            aString: {
                type: 'string'
            },
            aNumber: {
                type: 'number'
            },
            aBoolean: {
                type: 'boolean'
            },
            anArrayOfStrings: {
                type: 'array',
                items: {
                    type: 'string'
                }
            },
            anObject: {
                type: 'object',
                required: ['anotherString'],
                properties: {
                    anotherString: {
                        type: 'string'
                    }
                }
            },
            anArrayOfObjects: {
                type: 'array',
                items: {
                    type: 'object',
                    required: ['anotherString'],
                    properties: {
                        anotherString: {
                            type: 'string'
                        }
                    }
                }
            },
            anArrayOfReferencedModel: {
                type: 'array',
                items: {
                    $ref: '#/definitions/Address'
                }
            },
            aUnion: {
                type: ['string', 'number']
            }
        }
    };

    describe('nodeSyncSnippet', () => {
        it('should match snapshot', () => {
            const snippet = nodeSyncSnippet({
                modelName: 'users',
                secretKey: 'secret',
                connectionId: 'connectionId',
                providerConfigKey: 'providerConfigKey'
            });

            expect(snippet).toMatchSnapshot();
        });
    });

    describe('nodeActionSnippet', () => {
        it('should match snapshot', () => {
            const snippet = nodeActionSnippet({
                actionName: 'createUser',
                secretKey: 'secret',
                connectionId: 'connectionId',
                providerConfigKey: 'providerConfigKey',
                input: sampleInput
            });

            expect(snippet).toMatchSnapshot();
        });
    });

    describe('httpSnippet', () => {
        const baseTestData = {
            baseUrl: 'https://api.nango.dev',
            secretKey: 'nango_secret_key_123456789',
            connectionId: 'test_connection_id',
            providerConfigKey: 'github'
        };

        describe('GET endpoint', () => {
            const getEndpoint: NangoSyncEndpointV2 = {
                method: 'GET',
                path: '/users'
            };

            it('should generate cURL snippet', async () => {
                const snippet = await httpSnippet({
                    ...baseTestData,
                    endpoint: getEndpoint,
                    language: 'shell'
                });

                expect(snippet).toMatchSnapshot();
            });

            it('should generate Java snippet', async () => {
                const snippet = await httpSnippet({
                    ...baseTestData,
                    endpoint: getEndpoint,
                    language: 'java'
                });

                expect(snippet).toMatchSnapshot();
            });

            it('should generate Python snippet', async () => {
                const snippet = await httpSnippet({
                    ...baseTestData,
                    endpoint: getEndpoint,
                    language: 'python'
                });

                expect(snippet).toMatchSnapshot();
            });

            it('should generate Go snippet', async () => {
                const snippet = await httpSnippet({
                    ...baseTestData,
                    endpoint: getEndpoint,
                    language: 'go'
                });

                expect(snippet).toMatchSnapshot();
            });

            it('should generate PHP snippet', async () => {
                const snippet = await httpSnippet({
                    ...baseTestData,
                    endpoint: getEndpoint,
                    language: 'php'
                });

                expect(snippet).toMatchSnapshot();
            });
        });

        describe('POST endpoint with input', () => {
            const postEndpoint: NangoSyncEndpointV2 = {
                method: 'POST',
                path: '/users'
            };

            it('should generate cURL snippet with input', async () => {
                const snippet = await httpSnippet({
                    ...baseTestData,
                    endpoint: postEndpoint,
                    language: 'shell',
                    input: sampleInput
                });

                expect(snippet).toMatchSnapshot();
            });

            it('should generate Java snippet with input', async () => {
                const snippet = await httpSnippet({
                    ...baseTestData,
                    endpoint: postEndpoint,
                    language: 'java',
                    input: sampleInput
                });

                expect(snippet).toMatchSnapshot();
            });

            it('should generate Python snippet with input', async () => {
                const snippet = await httpSnippet({
                    ...baseTestData,
                    endpoint: postEndpoint,
                    language: 'python',
                    input: sampleInput
                });

                expect(snippet).toMatchSnapshot();
            });

            it('should generate Go snippet with input', async () => {
                const snippet = await httpSnippet({
                    ...baseTestData,
                    endpoint: postEndpoint,
                    language: 'go',
                    input: sampleInput
                });

                expect(snippet).toMatchSnapshot();
            });

            it('should generate PHP snippet with input', async () => {
                const snippet = await httpSnippet({
                    ...baseTestData,
                    endpoint: postEndpoint,
                    language: 'php',
                    input: sampleInput
                });

                expect(snippet).toMatchSnapshot();
            });
        });

        describe('with hideSecret option', () => {
            const getEndpoint: NangoSyncEndpointV2 = {
                method: 'GET',
                path: '/users'
            };

            it('should generate cURL snippet with hidden secret', async () => {
                const snippet = await httpSnippet({
                    ...baseTestData,
                    endpoint: getEndpoint,
                    language: 'shell',
                    hideSecret: true
                });

                expect(snippet).toMatchSnapshot();
            });

            it('should generate cURL snippet with visible secret', async () => {
                const snippet = await httpSnippet({
                    ...baseTestData,
                    endpoint: getEndpoint,
                    language: 'shell',
                    hideSecret: false
                });

                expect(snippet).toMatchSnapshot();
            });
        });
    });
});
