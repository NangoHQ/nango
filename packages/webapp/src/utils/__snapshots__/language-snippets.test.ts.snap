// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`language snippets > httpSnippet > GET endpoint > should generate Go snippet 1`] = `
"package main

import (
	"fmt"
	"net/http"
	"io"
)

func main() {

	url := "https://api.nango.dev/v1/users"

	req, err := http.NewRequest("GET", url, nil)

	if err != nil {
		panic(err)
	}
	req.Header.Add("Authorization", "Bearer nang****")
	req.Header.Add("Content-Type", "application/json")
	req.Header.Add("Connection-Id", "test_connection_id")
	req.Header.Add("Provider-Config-Key", "github")

	res, err := http.DefaultClient.Do(req)
	if err != nil {
		panic(err)
	}

	defer res.Body.Close()
	body, err := io.ReadAll(res.Body)
	if err != nil {
		panic(err)
	}

	fmt.Println(res)
	fmt.Println(string(body))

}"
`;

exports[`language snippets > httpSnippet > GET endpoint > should generate Java snippet 1`] = `
"HttpResponse<String> response = Unirest.get("https://api.nango.dev/v1/users")
  .header("Authorization", "Bearer nang****")
  .header("Content-Type", "application/json")
  .header("Connection-Id", "test_connection_id")
  .header("Provider-Config-Key", "github")
  .asString();"
`;

exports[`language snippets > httpSnippet > GET endpoint > should generate PHP snippet 1`] = `
"<?php

$client = new \\GuzzleHttp\\Client();

$response = $client->request('GET', 'https://api.nango.dev/v1/users', [
  'headers' => [
    'Authorization' => 'Bearer nang****',
    'Connection-Id' => 'test_connection_id',
    'Content-Type' => 'application/json',
    'Provider-Config-Key' => 'github',
  ],
]);

echo $response->getBody();"
`;

exports[`language snippets > httpSnippet > GET endpoint > should generate Python snippet 1`] = `
"import http.client

conn = http.client.HTTPSConnection("api.nango.dev")

headers = {
    'Authorization': "Bearer nang****",
    'Content-Type': "application/json",
    'Connection-Id': "test_connection_id",
    'Provider-Config-Key': "github"
}

conn.request("GET", "/v1/users", headers=headers)

res = conn.getresponse()
data = res.read()

print(data.decode("utf-8"))"
`;

exports[`language snippets > httpSnippet > GET endpoint > should generate cURL snippet 1`] = `
"curl --request GET \\
  --url https://api.nango.dev/v1/users \\
  --header 'Authorization: Bearer nang****' \\
  --header 'Connection-Id: test_connection_id' \\
  --header 'Content-Type: application/json' \\
  --header 'Provider-Config-Key: github'"
`;

exports[`language snippets > httpSnippet > POST endpoint with input > should generate Go snippet with input 1`] = `
"package main

import (
	"fmt"
	"strings"
	"net/http"
	"io"
)

func main() {

	url := "https://api.nango.dev/v1/users"

	payload := strings.NewReader("{ \\"aString\\": \\"<string>\\", \\"aNumber\\": \\"<number>\\", \\"aBoolean\\": \\"<boolean>\\", \\"anArrayOfStrings\\": \\"<string[]>\\", \\"anObject\\": { \\"anotherString\\": \\"<string>\\" }, \\"anArrayOfObjects\\": \\"<{ \\"anotherString\\": \\"<string>\\" }[]>\\", \\"anArrayOfReferencedModel\\": \\"<Address[]>\\", \\"aUnion\\": \\"<string | number>\\" }")

	req, err := http.NewRequest("POST", url, payload)

	if err != nil {
		panic(err)
	}
	req.Header.Add("Authorization", "Bearer nang****")
	req.Header.Add("Content-Type", "application/json")
	req.Header.Add("Connection-Id", "test_connection_id")
	req.Header.Add("Provider-Config-Key", "github")

	res, err := http.DefaultClient.Do(req)
	if err != nil {
		panic(err)
	}

	defer res.Body.Close()
	body, err := io.ReadAll(res.Body)
	if err != nil {
		panic(err)
	}

	fmt.Println(res)
	fmt.Println(string(body))

}"
`;

exports[`language snippets > httpSnippet > POST endpoint with input > should generate Java snippet with input 1`] = `
"HttpResponse<String> response = Unirest.post("https://api.nango.dev/v1/users")
  .header("Authorization", "Bearer nang****")
  .header("Content-Type", "application/json")
  .header("Connection-Id", "test_connection_id")
  .header("Provider-Config-Key", "github")
  .body("{ \\"aString\\": \\"<string>\\", \\"aNumber\\": \\"<number>\\", \\"aBoolean\\": \\"<boolean>\\", \\"anArrayOfStrings\\": \\"<string[]>\\", \\"anObject\\": { \\"anotherString\\": \\"<string>\\" }, \\"anArrayOfObjects\\": \\"<{ \\"anotherString\\": \\"<string>\\" }[]>\\", \\"anArrayOfReferencedModel\\": \\"<Address[]>\\", \\"aUnion\\": \\"<string | number>\\" }")
  .asString();"
`;

exports[`language snippets > httpSnippet > POST endpoint with input > should generate PHP snippet with input 1`] = `
"<?php

$client = new \\GuzzleHttp\\Client();

$response = $client->request('POST', 'https://api.nango.dev/v1/users', [
  'body' => '{ "aString": "<string>", "aNumber": "<number>", "aBoolean": "<boolean>", "anArrayOfStrings": "<string[]>", "anObject": { "anotherString": "<string>" }, "anArrayOfObjects": "<{ "anotherString": "<string>" }[]>", "anArrayOfReferencedModel": "<Address[]>", "aUnion": "<string | number>" }',
  'headers' => [
    'Authorization' => 'Bearer nang****',
    'Connection-Id' => 'test_connection_id',
    'Content-Type' => 'application/json',
    'Provider-Config-Key' => 'github',
  ],
]);

echo $response->getBody();"
`;

exports[`language snippets > httpSnippet > POST endpoint with input > should generate Python snippet with input 1`] = `
"import http.client

conn = http.client.HTTPSConnection("api.nango.dev")

payload = "{ \\"aString\\": \\"<string>\\", \\"aNumber\\": \\"<number>\\", \\"aBoolean\\": \\"<boolean>\\", \\"anArrayOfStrings\\": \\"<string[]>\\", \\"anObject\\": { \\"anotherString\\": \\"<string>\\" }, \\"anArrayOfObjects\\": \\"<{ \\"anotherString\\": \\"<string>\\" }[]>\\", \\"anArrayOfReferencedModel\\": \\"<Address[]>\\", \\"aUnion\\": \\"<string | number>\\" }"

headers = {
    'Authorization': "Bearer nang****",
    'Content-Type': "application/json",
    'Connection-Id': "test_connection_id",
    'Provider-Config-Key': "github"
}

conn.request("POST", "/v1/users", payload, headers)

res = conn.getresponse()
data = res.read()

print(data.decode("utf-8"))"
`;

exports[`language snippets > httpSnippet > POST endpoint with input > should generate cURL snippet with input 1`] = `
"curl --request POST \\
  --url https://api.nango.dev/v1/users \\
  --header 'Authorization: Bearer nang****' \\
  --header 'Connection-Id: test_connection_id' \\
  --header 'Content-Type: application/json' \\
  --header 'Provider-Config-Key: github' \\
  --data '{
    "aString": "<string>",
    "aNumber": "<number>",
    "aBoolean": "<boolean>",
    "anArrayOfStrings": "<string[]>",
    "anObject": { "anotherString": "<string>" },
    "anArrayOfObjects": "<{ "anotherString": "<string>" }[]>",
    "anArrayOfReferencedModel": "<Address[]>",
    "aUnion": "<string | number>"
}'"
`;

exports[`language snippets > httpSnippet > with hideSecret option > should generate cURL snippet with hidden secret 1`] = `
"curl --request GET \\
  --url https://api.nango.dev/v1/users \\
  --header 'Authorization: Bearer nang****' \\
  --header 'Connection-Id: test_connection_id' \\
  --header 'Content-Type: application/json' \\
  --header 'Provider-Config-Key: github'"
`;

exports[`language snippets > httpSnippet > with hideSecret option > should generate cURL snippet with visible secret 1`] = `
"curl --request GET \\
  --url https://api.nango.dev/v1/users \\
  --header 'Authorization: Bearer nango_secret_key_123456789' \\
  --header 'Connection-Id: test_connection_id' \\
  --header 'Content-Type: application/json' \\
  --header 'Provider-Config-Key: github'"
`;

exports[`language snippets > nodeActionSnippet > should match snapshot 1`] = `
"import Nango from '@nangohq/node';
const nango = new Nango({ secretKey: 'secr****' });
const response = await nango.triggerAction(
    'providerConfigKey',
    'connectionId',
    'createUser',
    {
        "aString": "<string>",
        "aNumber": "<number>",
        "aBoolean": "<boolean>",
        "anArrayOfStrings": "<string[]>",
        "anObject": { "anotherString": "<string>" },
        "anArrayOfObjects": "<{ "anotherString": "<string>" }[]>",
        "anArrayOfReferencedModel": "<Address[]>",
        "aUnion": "<string | number>"
    }
);"
`;

exports[`language snippets > nodeSyncSnippet > should match snapshot 1`] = `
"import { Nango } from '@nangohq/node';
const nango = new Nango({ secretKey: 'secr****' });

const records = await nango.listRecords({
    providerConfigKey: 'providerConfigKey',
    connectionId: 'connectionId',
    model: 'users'
});"
`;
